### ISSUE TITLE
Getting ScaleGMN to classify/train on larger SIRENs

### ISSUE THREAD

**Issue by JLuij** [jkalogero/scalegmn#3]
Hi,

I am trying to use ScaleGMN to classify a set of SIRENs but I'm struggling to get it to learn on even a small subset of the data. My SIREN architecture is [2,128,128,128,1] which is relatively large but necessary to fit the high-resolution images I'm using. All SIRENs were initialised with identical weights. Canonicalisation did not affect the parameters.

Even though I've set the debug flag to train on just 16 SIRENs, after 100 epochs the train accuracy has only peaked at 37.5%. The dataset code was adjusted to load the weights (as the layers have different names and weight order), and I've changed the config somewhat (see below).

Do you have any suggestions to make the data and ScaleGMN compatible, perhaps something I've overlooked?

Thank you!

```
batch_size: 16
data:
  dataset: labeled_fashion_mnist_inr
  dataset_path: /path/to/dataset
  split_path: /path/to/dataset/mnist_splits.json
  image_size: [1024,1024]
  node_pos_embed: &node_pos_embed True
  edge_pos_embed: &edge_pos_embed True
  # the below can be extracted per datapoint, but since it is the same for all, we can define it here
  layer_layout: [2, 128, 128, 128, 1]
  switch_to_canon: False # asdfa

train_args:
  num_epochs: 750
  val_acc_threshold: 0.1
  patience: 50
  seed: -1
  loss: CrossEntropyLoss


scalegmn_args:
  d_in_v: &d_in_v 1  # initial dimension of input nn bias
  d_in_e: &d_in_e 1  # initial dimension of input nn weights
  d_hid: &d_hid 128  # hidden dimension
  num_layers: 4 # number of gnn layers to apply
  direction: forward
  equivariant: False
  symmetry: sign  # symmetry
  jit: False # prefer compile - compile gnn to optimize performance
  compile: False # compile gnn to optimize performance

  readout_range: full_graph  # or full_graph
  gnn_skip_connections: True

  concat_mlp_directions: False  # only for bidirectional: apply an MLP before concatenating the forward and backward directions
  reciprocal: True  # only for bidirectional

  node_pos_embed: *node_pos_embed  # use positional encodings
  edge_pos_embed: *edge_pos_embed  # use positional encodings

  graph_init:
    d_in_v: *d_in_v
    d_in_e: *d_in_e
    project_node_feats: True
    project_edge_feats: True
    d_node: *d_hid
    d_edge: *d_hid

  positional_encodings:
    final_linear_pos_embed: False
    sum_pos_enc: False
    po_as_different_linear: False
    equiv_net: False
    # args for the equiv net option.
    sum_on_io: True
    equiv_on_hidden: True
    num_mlps: 3
    layer_equiv_on_hidden: False

  gnn_args:
    d_hid: *d_hid
    message_fn_layers: 1
    message_fn_skip_connections: False
    update_node_feats_fn_layers: 1
    update_node_feats_fn_skip_connections: False
    update_edge_attr: True
    dropout: 0.2
    dropout_all: True  # False: only in between the gnn layers, True: + all mlp layers
    update_as_act: False
    update_as_act_arg: sum
    mlp_on_io: True

    msg_equiv_on_hidden: True
    upd_equiv_on_hidden: True
    layer_msg_equiv_on_hidden: False
    layer_upd_equiv_on_hidden: False
    msg_num_mlps: 3
    upd_num_mlps: 3
    pos_embed_msg: False
    pos_embed_upd: False
    layer_norm: False
    aggregator: add
    sign_symmetrization: True

  mlp_args:
    d_k: [64]
    activation: silu
    dropout: 0.
    final_activation: identity
    batch_norm: False  # check again
    layer_norm: True
    bias: True
    skip: False

  readout_args:
    d_out: 10  # final dimension of readout Permutation Invariant Scale(Sign)Net - Number of Classes for classification tasks.
    d_rho: 64  # intermediate dimension within Readout module


optimization:
  clip_grad: True
  clip_grad_max_norm: 10.0
  optimizer_name: AdamW
  optimizer_args:
    lr: 5e-3
    weight_decay: 0.001
  scheduler_args:
    scheduler: WarmupLRScheduler
    warmup_steps: 10
    scheduler_mode: min
    decay_rate: 0
    decay_steps: 0
    patience: None
    min_lr: None

wandb_args:
  project: fmnist_cls
  entity: null
  group: null
  name: testing
  tags: null
```

**Comment by jkalogero** [comment]
Hello, 

Sorry for not getting back to you sooner.

Assuming the INRs are fitted properly to the original grayscale images, there are a couple of things you could check:

- correctly parsing of the input network
- graph is constructed as expected (number of nodes/edges and positional encodings)
- your INRs have the classic SIREN architecture (no changes that could affect the exhibited symmetries)

Regarding hyperparameters, try experimenting with:
- d_hid
- num_layers
- mlp_args.layer_norm
- gnn_skip_connections
- dropout(_all)
- node_pos_embed, edge_pos_embed

What is the training accuracy when trained on the whole training split?
As shown in our experiments, you could achieve better results when augmenting the training split with extra _views_, i.e., the same original image, but different initialization. If everything else fails you could also try this.

Unfortunately, I am not sure how else I can assist you. In any case, you should keep in mind that current works on metanetworks have been primarily evaluated on "toy" datasets, hence it is expected that they may struggle with significantly more challenging datasets.

Let me know if any of these helped.
